Update for IETF 122 (#1)
rohanmahy/mahy-mls-xwing: draft-mahy-mls-pq.md

---
title: ML-KEM and Hybrid Cipher Suites for Messaging Layer Security

> abbrev: MLS with ML-KEM

Author: Disagree.  Can't use ML-KEM without a ciphersuite.

> abbrev: MLS with ML-KEM

Reviewer: I don't have a strong feeling here, but I think a lot of readers will see ML-KEM in this way and assume this is the entire MLS PQ strategy. I would spell it out to avoid confusion, not because what you wrote is inaccurate.

workgroup: MLS
area: sec
category: info
keyword:
 - ML-KEM
 - Kyber
 - post-quantum
 - post quantum KEM
 - KEM
 - PQ/T
 - hybrid
 - hybrid KEM
 - Key Exchange Mechanism

stand_alone: yes
pi: [toc, sortrefs, symrefs]

venue:
  group: MLS
  type: Working Group
  mail: mls@ietf.org
  arch: https://mailarchive.ietf.org/arch/browse/mls/
  github: rohanmahy/mahy-mls-xwing/
#  latest: https://github.com/rohanmahy/mahy-mls-xwing/latest

author:
 -  ins: R. Mahy
    name: Rohan Mahy
    organization: Unaffiliated
    email: rohan.ietf@gmail.com
 -  ins: R.L. Barnes
    name: Richard L. Barnes
    organization: Cisco
    email: rlb@ipv.sx

normative:
  SHS: DOI.10.6028/NIST.FIPS.180-4
  MLKEM: DOI.10.6028/NIST.FIPS.203

informative:

--- abstract

This document reigsters new cipher suites for Messaging Layer Security (MLS)
based on "post-quantum" algorithms, which are intended to be resilient to attack
by quantum computers.  These cipher suites are constructed using the new
Module-Lattice Key Encapsulation Mechanism (ML-KEM), optionally in combination
with traditional elliptic curve KEMs, together with appropriate authenticated

> encryption and hash algorithms.

Author: ```suggestion
encryption, hash, and signature algorithms.
```

--- middle

# Introduction

The potential availability of a cryptographically-relevant quantum
computer has caused concern that well-funded adversaries could overturn
long-held assumptions about the security assurances of classical
Key Exchange Mechanisms (KEMs) and classical cryptographic signatures,
which are fundamental to modern security protocols, including the MLS
protocol {{!RFC9420}}.

Of particular concern are "harvest now, decrypt later" attacks, by which an
attacker could collect encrypted traffic now, before a quantum computer exists,
and later use a quantum computer to break the confidentiality protections on the
collected traffic.

In response to these concerns, the cryptographic community has defined
"post-quantum" algorithms, which are designed to be resilient to attacks by
quantum computers.  Symmetric algorithms can be made post-quantum secure simply

> by using longer keys and hashes.  For asymmetric operations such as KEM and

Author: Disagree.  For example, CNSA 2.0 forbids AES-128.

> by using longer keys and hashes.  For asymmetric operations such as KEM and

Reviewer: ok, AES256 with the CNSA consistency and Grover's

pure ML-KEM:

* ML-KEM-768 + X25519 (Medium security, Non-NIST, PQ/T hybrid)
* ML-KEM-768 + P-256 (Medium security, NIST, PQ/T hybrid)
* ML-KEM-1024 + P-384 (High security, NIST, PQ/T hybrid)
* ML-KEM-768 (Medium security, NIST, pure PQ)
* ML-KEM-1024 (High security, NIST, pure PQ)

For the PQ/T hybrid cipher suites, we use the KEM combinators defined in
{{!I-D.connolly-cfrg-xwing-kem}} and {{!I-D.irtf-cfrg-hybrid-kems}}.  For the
pure-PQ cipher suites, we use the HPKE integration for ML-KEM defined in
{{!I-D.connolly-cfrg-hpke-mlkem}}.

# IANA Considereations

This document requests that IANA add the following entries to the "MLS Cipher
Suites" registry, replacing "XXXX" with the RFC number assigned to this document:

| Value  | Name                                       | Rec | Reference |
|:=======|:===========================================|:===:|:=========:|
| 0xTBD1 | MLS_128_X_Wing_AES256GCM_SHA512_Ed25519    |  Y  | [RFCXXXX] |
| 0xTBD2 | MLS_128_QSF_768_AES256GCM_SHA512_P256      |  Y  | [RFCXXXX] |
| 0xTBD3 | MLS_256_QSF_1024_AES256GCM_SHA512_P384     |  Y  | [RFCXXXX] |
| 0xTBD4 | MLS_128_ML_KEM_768_AES256GCM_SHA512_P256   |  Y  | [RFCXXXX] |

> | 0xTBD5 | MLS_256_ML_KEM_1024_AES256GCM_SHA512_P384  |  Y  | [RFCXXXX] |

Reviewer: I don't these MLS combos make sense, and the AFAIK the names are wrong

There is no reason to use SHA512 for everything. Only the P521 and Ed448 MLS ciphersuites currently use SHA512. Align the security levels and pick a matching HASH and AEAD (see exception **)
Likewise, for the ciphersuites that use already need SHA3 in their KDF, maybe we should use SHA3 as the MLS ciphersuite HASH function?
(**) CSNA 2.0 sort of deprecated SHA256 even for the 128 bit security level; people who want to use pure PQ are almost certainly following CSNA and therefore probably need SHA384.

MLS_128_X-Wing_AES128GCM_SHA256_Ed25519
MLS_128_QSF-KEM(ML-KEM-768,P-256)-XOF(SHAKE256)-KDF(SHA3-256)_AES128GCM_SHA3-256_P256
MLS_192_QSF-KEM(ML-KEM-1024,P-384)-XOF(SHAKE256)-KDF(SHA3-256)_AES256GCM_SHA3-384_P384
MLS_128_ML-KEM-768_AES128GCM_SHA384_P256 (**)
MLS_192_ML-KEM-1024_AES256GCM_SHA384_P384

Reviewer: @Dierdre , do you have an opinion of SHA3 vs SHA2 for the MLS ciphersuite hash function for anyone doing QFS?

Author: * The thinking on AES-256 and SHA-512 was motivated by the "double the key/hash size" approach to PQ for symmetric.
* I don't have as strong feeling about SHA2 vs. SHA3
    * On the one hand, yes, folks will have SHA3 for other reasons in some cases
    * On the other hand, CNSA 2.0 still lists SHA2 as "General Purpose" and SHA3 as "In Specific Applications"
    * In any case, CNSA 2.0 does forbid the 256-bit hash length
* Using the full QSF expansions seems really unwieldy, and unnecessary because these are just mnemonics; the specific algorithms are detailed below.
* This is just an initial stab, we can debate in the WG

Counter-offer: 

```
MLS_128_X_Wing_AES256GCM_SHA512_Ed25519
MLS_128_QSF_MLKEM768_P256_AES256GCM_SHA3_512_P256
MLS_192_QSF-MLKEM768_P256_AES256GCM_SHA3_512_P384
MLS_128_MLKEM768_AES128GCM_SHA512_P256
MLS_192_MLKEM1024_AES256GCM_SHA512_P384
```

Reviewer 2: > @dconnolly , do you have an opinion of SHA3 vs SHA2 for the MLS ciphersuite hash function for anyone doing QFS?

SHA3 is a secure KDF, SHA2 used on its own outside of HKDF is not

Reviewer 2: > @dconnolly , do you have an opinion of SHA3 vs SHA2 for the MLS ciphersuite hash function for anyone doing QFS?

Oh I have to look up again how the MLS ciphersuite uses its parameterized hash function outside of the hybrid KEM context, which needs a secure KDF

Reviewer: > > @dconnolly , do you have an opinion of SHA3 vs SHA2 for the MLS ciphersuite hash function for anyone doing QFS?
> 
> Oh I have to look up again how the MLS ciphersuite uses its parameterized hash function outside of the hybrid KEM context, which needs a secure KDF

@dconnolly 
MLS uses SHA256/384/512 in all currently-registered ciphersuites for generating KeyPackageRefs, ProposalRefs, membership_tag, confirmation_hash, tree_hash, etc.

MLS use key derivation extensively in the Key Schedule.
Each existing MLS ciphersuite has an **_implicit_** HPKE KDF algorithm as shown in [RFC 9420 Table 7](https://www.rfc-editor.org/rfc/rfc9420.html#table-7). The only IANA registrations currently are HKDF-SHA256/384/512. This is clearly **NOT**  what we want for the combiners and for plain ML-KEM. For one thing the key length is wrong if we use anything except HKDF-SHA256, and all the ML-KEM implementations will have a SHAKE library somewhere. I don't understand the combiners well enough to understand if you can just IANA register SHAKE256 as an HPKE KDF, or if all three algorithm documents (XWing, irtf-cfrg-hybrid-kems, connolly-cfrg-hpke-mlkem) should also be IANA registering one or more HPKE KDFs that are particular for their key derivation procedures.

Reviewer: OK, I think draft-irtf-cfrg-hybrid-kems needs to IANA register both `KDF(SHA3-256)` (referencing Section 6.1 of FIPS202(, and `XOF(SHAKE256)` (referencing Section 6.2 of FIPS202) in the IANA HPKE KDF registry. Anyone?

> > > @dconnolly , do you have an opinion of SHA3 vs SHA2 for the MLS ciphersuite hash function for anyone doing QFS?
> > 
> > 
> > Oh I have to look up again how the MLS ciphersuite uses its parameterized hash function outside of the hybrid KEM context, which needs a secure KDF
> 
> @dconnolly MLS uses SHA256/384/512 in all currently-registered ciphersuites for generating KeyPackageRefs, ProposalRefs, membership_tag, confirmation_hash, tree_hash, etc.
> 
> MLS use key derivation extensively in the Key Schedule. Each existing MLS ciphersuite has an **_implicit_** HPKE KDF algorithm as shown in [RFC 9420 Table 7](https://www.rfc-editor.org/rfc/rfc9420.html#table-7). The only IANA registrations currently are HKDF-SHA256/384/512. This is clearly **NOT** what we want for the combiners and for plain ML-KEM. For one thing the key length is wrong if we use anything except HKDF-SHA256, and all the ML-KEM implementations will have a SHAKE library somewhere. I don't understand the combiners well enough to understand if you can just IANA register SHAKE256 as an HPKE KDF, or if all three algorithm documents (XWing, irtf-cfrg-hybrid-kems, connolly-cfrg-hpke-mlkem) should also be IANA registering one or more HPKE KDFs that are particular for their key derivation procedures.

> | 0xTBD1 | MLS_128_X_Wing_AES256GCM_SHA512_Ed25519    |  Y  | [RFCXXXX] |
> | 0xTBD2 | MLS_128_QSF_768_AES256GCM_SHA512_P256      |  Y  | [RFCXXXX] |
> | 0xTBD3 | MLS_256_QSF_1024_AES256GCM_SHA512_P384     |  Y  | [RFCXXXX] |
> | 0xTBD4 | MLS_128_ML_KEM_768_AES256GCM_SHA512_P256   |  Y  | [RFCXXXX] |
> | 0xTBD5 | MLS_256_ML_KEM_1024_AES256GCM_SHA512_P384  |  Y  | [RFCXXXX] |

Reviewer: ```suggestion
| TBD1 | MLS_128_X_Wing_AES256GCM_SHA384_Ed25519    |  Y  | [RFCXXXX] |
| TBD2 | MLS_128_QSF-KEM(ML-KEM-768,P-256)-0xTBDH0_AES256GCM_SHA384_P256      |  Y  | [RFCXXXX] |
| TBD3 | MLS_192_QSF-KEM(ML-KEM-1024,P-384)-0xTBDH1_AES256GCM_SHA384_P384     |  Y  | [RFCXXXX] |
| TBD4 | MLS_128_ML_KEM_768_AES256GCM_SHA384_P256   |  Y  | [RFCXXXX] |
| TBD5 | MLS_192_ML_KEM_1024_AES256GCM_SHA384_P384  |  Y  | [RFCXXXX] |

> Note to IANA, please replace `THDH0` and `TBDH1` in lines 2 and 3 of the table.
```

All of these cipher suites use HMAC [RFC2104] with SHA512 as their MAC function.
The mapping of cipher suites to HPKE primitives [RFC9180], HMAC hash functions,
and TLS signature schemes [RFC8446] is as follows:

| Value  | KEM     | KDF    | AEAD   | Hash   | Signature              |
|:=======|:========|:=======|:=======|:=======|:=======================|

> | 0xTBD1 | 0x647a  | 0x0003 | 0x0002 | SHA512 | ed25519                |
> | 0xTBD2 | 0xTBDH0 | 0x0003 | 0x0002 | SHA512 | ecdsa_secp256r1_sha256 |
> | 0xTBD3 | 0xTBDH1 | 0x0003 | 0x0002 | SHA512 | ecdsa_secp384r1_sha384 |
> | 0xTBD4 | 0x0041  | 0x0003 | 0x0002 | SHA512 | ecdsa_secp256r1_sha256 |
> | 0xTBD5 | 0x0042  | 0x0003 | 0x0002 | SHA512 | ecdsa_secp384r1_sha384 |

Reviewer: ```suggestion
| 0xTBD1 | 0x647a  | 0x0002 | 0x0002 | SHA384 | ed25519                |
| 0xTBD2 | 0xTBDH0 | 0x0002 | 0x0002 | SHA384 | ecdsa_secp256r1_sha256 |
| 0xTBD3 | 0xTBDH1 | 0x0002 | 0x0002 | SHA384 | ecdsa_secp384r1_sha384 |
| 0xTBD4 | 0x0041  | 0x0002 | 0x0002 | SHA384 | ecdsa_secp256r1_sha256 |
| 0xTBD5 | 0x0042  | 0x0002 | 0x0002 | SHA384 | ecdsa_secp384r1_sha384 |
```

> The values `0xTBDH[01]` refer to the code points to be assigned by IANA for

Reviewer: ```suggestion
The values `TBDH0` and `TBDH1` refer to the code points to be assigned by IANA for
```

the following hybrid KEMs defined in {{!I-D.irtf-cfrg-hybrid-kems}}:

> * `0xTBDH0 = QSF-KEM(ML-KEM-768,P-256)-XOF(SHAKE256)-KDF(SHA3-256)`
> * `0xTBDH1 = QSF-KEM(ML-KEM-1024,P-384)-XOF(SHAKE256)-KDF(SHA3-256)`

Reviewer: ```suggestion
* `TBDH0 = QSF-KEM(ML-KEM-768,P-256)-XOF(SHAKE256)-KDF(SHA3-256)`
* `TBDH1 = QSF-KEM(ML-KEM-1024,P-384)-XOF(SHAKE256)-KDF(SHA3-256)`
```
